### ISSUE TITLE
kcov (v30) does not find debugging symbols, but nm and gdb says there are (nix-related?)

### ISSUE THREAD

**Issue by FlorentBecker** [SimonKagstrom/kcov#123]
When I try to run some rust executable (hello world) through kcov, kcov complains that it cannot find debugging symbols:
```
$ kcov --version
kcov 30
$ kcov --debug 4 target/cov/ target/debug/trivial-2b7adf9f4183aea6
kcov --debug 4 target/cov/ target/debug/trivial-2b7adf9f4183aea6 
No debug symbols in target/debug/trivial-2b7adf9f4183aea6.
Cannot open linux-vdso.so.1
[...]
```

Yet, both nm and gdb are able to find the debugging symbols:
```
$ nm target/debug/trivial-2b7adf9f4183aea6  | wc -l
3487
$ gdb target/debug/trivial-2b7adf9f4183aea6 
GNU gdb (GDB) 7.10
Copyright (C) 2015 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-unknown-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from target/debug/trivial-2b7adf9f4183aea6...done.
warning: Missing auto-load script at offset 0 in section .debug_gdb_scripts
of file /home/florent/projets_perso/rust_triv/trivial/target/debug/trivial-2b7adf9f4183aea6.
Use `info auto-load python-scripts [REGEXP]' to list them.
(gdb)
```

Of course, kcov then reports NaN coverage on 0 lines of code. Do you have any idea what can be happenning?

**Comment by SimonKagstrom** [comment]
That lists symbols, but does it have DWARF debug sections? You can check
that with readelf -S <binary>.

**Comment by FlorentBecker** [comment]
```
$ LANG=C readelf -S target/debug/trivial-2b7adf9f4183aea6 
There are 40 section headers, starting at offset 0xbd848:

Section Headers:
  [Nr] Name              Type             Address           Offset
       Size              EntSize          Flags  Link  Info  Align
  [ 0]                   NULL             0000000000000000  00000000
       0000000000000000  0000000000000000           0     0     0
  [ 1] .interp           PROGBITS         0000000000000270  00000270
       0000000000000050  0000000000000000   A       0     0     1
  [ 2] .note.ABI-tag     NOTE             00000000000002c0  000002c0
       0000000000000020  0000000000000000   A       0     0     4
  [ 3] .hash             HASH             00000000000002e0  000002e0
       00000000000006bc  0000000000000004   A       4     0     8
  [ 4] .dynsym           DYNSYM           00000000000009a0  000009a0
       00000000000015c0  0000000000000018   A       5     3     8
  [ 5] .dynstr           STRTAB           0000000000001f60  00001f60
       0000000000002e23  0000000000000000   A       0     0     1
  [ 6] .gnu.version      VERSYM           0000000000004d84  00004d84
       00000000000001d0  0000000000000002   A       4     0     2
  [ 7] .gnu.version_r    VERNEED          0000000000004f58  00004f58
       0000000000000160  0000000000000000   A       5     7     8
  [ 8] .rela.dyn         RELA             00000000000050b8  000050b8
       000000000000c258  0000000000000018   A       4     0     8
  [ 9] .rela.plt         RELA             0000000000011310  00011310
       00000000000008a0  0000000000000018   A       4    11     8
  [10] .init             PROGBITS         0000000000011bb0  00011bb0
       000000000000001a  0000000000000000  AX       0     0     4
  [11] .plt              PROGBITS         0000000000011bd0  00011bd0
       00000000000005d0  0000000000000010  AX       0     0     16
  [12] .text             PROGBITS         00000000000121a0  000121a0
       0000000000081ea0  0000000000000000  AX       0     0     16
  [13] .fini             PROGBITS         0000000000094040  00094040
       0000000000000009  0000000000000000  AX       0     0     4
  [14] .rodata           PROGBITS         0000000000094080  00094080
       000000000000b0af  0000000000000000   A       0     0     64
  [15] .debug_gdb_script PROGBITS         000000000009f12f  0009f12f
       0000000000000022  0000000000000000 AMS       0     0     1
  [16] .eh_frame_hdr     PROGBITS         000000000009f154  0009f154
       0000000000001b2c  0000000000000000   A       0     0     4
  [17] .eh_frame         PROGBITS         00000000000a0c80  000a0c80
       0000000000008de4  0000000000000000   A       0     0     8
  [18] .gcc_except_table PROGBITS         00000000000a9a64  000a9a64
       0000000000009228  0000000000000000   A       0     0     4
  [19] .tdata            PROGBITS         00000000002b3000  000b3000
       00000000000000e8  0000000000000000 WAT       0     0     16
  [20] .init_array       INIT_ARRAY       00000000002b30e8  000b30e8
       0000000000000010  0000000000000000  WA       0     0     8
  [21] .fini_array       FINI_ARRAY       00000000002b30f8  000b30f8
       0000000000000008  0000000000000000  WA       0     0     8
  [22] .jcr              PROGBITS         00000000002b3100  000b3100
       0000000000000008  0000000000000000  WA       0     0     8
  [23] .data.rel.ro      PROGBITS         00000000002b3140  000b3140
       0000000000008a00  0000000000000000  WA       0     0     64
  [24] .dynamic          DYNAMIC          00000000002bbb40  000bbb40
       0000000000000250  0000000000000010  WA       5     0     8
  [25] .got              PROGBITS         00000000002bbd90  000bbd90
       0000000000000560  0000000000000008  WA       0     0     8
  [26] .got.plt          PROGBITS         00000000002bc2f0  000bc2f0
       00000000000002f8  0000000000000008  WA       0     0     8
  [27] .data             PROGBITS         00000000002bc600  000bc600
       0000000000000289  0000000000000000  WA       0     0     32
  [28] .bss              NOBITS           00000000002bc8c0  000bc889
       0000000000001280  0000000000000000  WA       0     0     64
  [29] .comment          PROGBITS         0000000000000000  000bc889
       0000000000000022  0000000000000001  MS       0     0     1
  [30] .debug_pubnames   PROGBITS         0000000000000000  000bc8ab
       00000000000000a2  0000000000000000           0     0     1
  [31] .debug_info       PROGBITS         0000000000000000  000bc94d
       00000000000004d7  0000000000000000           0     0     1
  [32] .debug_abbrev     PROGBITS         0000000000000000  000bce24
       00000000000000d3  0000000000000000           0     0     1
  [33] .debug_line       PROGBITS         0000000000000000  000bcef7
       0000000000000089  0000000000000000           0     0     1
  [34] .debug_str        PROGBITS         0000000000000000  000bcf80
       00000000000003d2  0000000000000001  MS       0     0     1
  [35] .debug_pubtypes   PROGBITS         0000000000000000  000bd352
       0000000000000339  0000000000000000           0     0     1
  [36] .debug_ranges     PROGBITS         0000000000000000  000bd68b
       0000000000000030  0000000000000000           0     0     1
  [37] .shstrtab         STRTAB           0000000000000000  000bd6bb
       0000000000000189  0000000000000000           0     0     1
  [38] .symtab           SYMTAB           0000000000000000  000be248
       0000000000014d48  0000000000000018          39   3089     8
  [39] .strtab           STRTAB           0000000000000000  000d2f90
       0000000000016052  0000000000000000           0     0     1
Key to Flags:
  W (write), A (alloc), X (execute), M (merge), S (strings), l (large)
  I (info), L (link order), G (group), T (TLS), E (exclude), x (unknown)
  O (extra OS processing required) o (OS specific), p (processor specific)
```

**Comment by SimonKagstrom** [comment]
On Sun, Nov 29, 2015 at 3:41 PM, FlorentBecker <notifications@github.com>
wrote:

> $ LANG=C readelf -S target/debug/trivial-2b7adf9f4183aea6
> There are 40 section headers, starting at offset 0xbd848:
>
> Section Headers:
>        00000000000000d3  0000000000000000           0     0     1
>   [33] .debug_line       PROGBITS         0000000000000000  000bcef7
>        0000000000000089  0000000000000000           0     0     1
>   [34] .debug_str        PROGBITS         0000000000000000  000bcf80
>        00000000000003d2  0000000000000001  MS       0     0     1
>   [35] .debug_pubtypes   PROGBITS         0000000000000000  000bd352
>        0000000000000339  0000000000000000           0     0     1
>   [36] .debug_ranges     PROGBITS         0000000000000000  000bd68b
>        0000000000000030  0000000000000000           0     0     1
>
>
> Ok, so it has the DWARF info as well, good! Can you email me the binary (
simon.kagstrom@gmail.com) so that I can take a look at it. I don't
understand this particular problem.

Did you test with earlier kcov versions (e.g., git checkout v29 if you
cloned the git repo)? I don't think there has been much work done in the
elf-parser, but I know a lot of people are using it for rust programs, so
it should work.

**Comment by FlorentBecker** [comment]
I had the same problem with kcov v29 (I tried with it first, then v30)

**Comment by SimonKagstrom** [comment]
OK, thanks for testing.

kcov will output this message if libdw cannot open/access the dwarf sections in the file. If you mail me the binary, I'll see if I can find out something more about why it doesn't work with this program.

**Comment by FlorentBecker** [comment]
I have that problem with a trivial C executable too (see attachment), it
seems to be a problem with my environment.
--
Florent

On 30/11/2015 11:08, Simon Kagstrom wrote:
>
> OK, thanks for testing.
>
> kcov will output this message if libdw cannot open/access the dwarf
> sections in the file. If you mail me the binary, I'll see if I can
> find out something more about why it doesn't work with this program.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SimonKagstrom/kcov/issues/123#issuecomment-160585004>.
>


#include <stdio.h>

int main()
{
  printf("hello world!\n");
  return 0;
  printf("you cant't see me!\n");
}

**Comment by FlorentBecker** [comment]
sorry, github does not seem to like binary attachments in bug reports.

**Comment by SimonKagstrom** [comment]
I've tested your emailed binary, and while I can't run the program, the DWARF parsing seems to work. I needed to use --report-only since the program can't start, but I get source/line -> address translations:

```
kcov --report-only --debug=15 /tmp/kcov ~/Downloads/hello 
BP registered at 0x400698
REPORT /tmp/hello.c:4 at 0x400698
BP registered at 0x40069c
REPORT /tmp/hello.c:5 at 0x40069c
BP registered at 0x4006a6
REPORT /tmp/hello.c:6 at 0x4006a6
BP registered at 0x4006ab
REPORT /tmp/hello.c:8 at 0x4006ab
BP registered at 0x4006ad
REPORT /tmp/hello.c:8 at 0x4006ad
```
could it be something with your libdw installation?

I get the following error when running the application btw:

```
~/Downloads/hello 
bash: /home/ska/Downloads/hello: /nix/store/y8scgsh8warhiy3pl39mqxyxphybcy7p-glibc-2.21/lib/ld-linux-x86-64.so.2: bad ELF interpreter: No such file or directory
```

**Comment by FlorentBecker** [comment]
I'm installing kcov using nix (nixos.org), hence the strange rpath you
see (and the error you get when running my example)

> could it be something with your libdw installation?

I've checked (using strace), and kcov does find libdw.so.1:

Konsole output
open("/nix/store/c9bi6s7mayv6blgfd2451ff2gvd9zz01-elfutils-0.163/lib/libdw.so.1",
O_RDONLY|O_CL
OEXEC) = 3

I'll check if a minimal example using libdw works on my machine.

**Comment by FlorentBecker** [comment]
Do you happen to have a minimal example of code that does the equivalent of `readelf -S` through libdw?

**Comment by SimonKagstrom** [comment]
Probably something like

```
int fd = open(filename, O_RDONLY, 0);
Dwarf *dbg = dwarf_begin(fd, DWARF_C_READ);
if (dbg)
   printf("Fine!");
else
   printf("Suspicious!\n");
```

should do. The code in kcov which handles this is in src/parsers/elf-parser.cc (parseOneDwarf()). You could also add some printouts there to see what fails. I suspect the problem occurs direcly after dwarf_begin, but there is also handling of build-id:s and debuglinks, which will be opened elsewhere.

**Comment by FlorentBecker** [comment]
Your code does print "fine!", I'll investigate further starting with src/parsers/elf-parser.cc

**Comment by FlorentBecker** [comment]
Strange: the following prints "Fine!" on a test executable:
```
#include <stdio.h>
#include <elfutils/libdw.h>

#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>

int main(int argc, char **argv)
{
int fd = open(argv[1], O_RDONLY, 0);
Dwarf *dbg = dwarf_begin(fd, DWARF_C_READ);
if (dbg)
   printf("Fine!");
else
   printf("Suspicious!\n");
}
```

Yet, running kcov on the same executable, dwarf_begin returns 0:

```
elf-parser.cc : 313
		/* Initialize libdwarf */
		dbg = dwarf_begin(fd, DWARF_C_READ);
                std::cout << "dwarf status of " << m_filename << ": " << dbg << std::endl;
```

prints `dwarf status of ./toto: 0`

**Comment by SimonKagstrom** [comment]
Oh, that *is* strange! Interesting problem!

Can you check with ldd if kcov really uses the same libdw as your example binary? I can't really explain this in any other way.

I'm not sure, but perhaps dwarf_begin sets errno? In that case, you could zero it and print it out after dwarf_begin, i.e.,

```
errno = 0;
dbg = dwarf_begin(...);
std::cout << ... << errno;
```

**Comment by SimonKagstrom** [comment]
If you track the kcov git HEAD, I've just refactored the code a bit (for Issue #42) and moved the DWARF stuff into a helper file, src/parsers/dwarf.cc. Should not affect this problem though!

**Comment by FlorentBecker** [comment]
This is getting stranger and stranger:
```
$ ldd dw_test | grep libdw
        libdw.so.1 => /nix/store/c9bi6s7mayv6blgfd2451ff2gvd9zz01-elfutils-0.163/lib/libdw.so.1 (0x00007fac665d4000)
$ ./dw_test hello
Fine!
$ ldd `which kcov` | grep libdw
        libdw.so.1 => /nix/store/c9bi6s7mayv6blgfd2451ff2gvd9zz01-elfutils-0.163/lib/libdw.so.1 (0x00007f32176bb000)
$ kcov cov/ hello
dwarf_status of hello:0with errno:0
[…]
```
(with git HEAD).

**Comment by SimonKagstrom** [comment]
To see if this has something to do with the build system after all, could you add the dwarf_begin() test stuff to src/main.cc before the rest of kcov starts. I.e., just open argv[2] and pass it into dwarf_begin().

If that is OK at that point, then kcov must have done something to break things afterwards.


Anyway, another thing I discovered is that libdw has a dwarf_errno() function - the return value of that would be interesting to print out instead of errno above.

Thanks for looking into this!

**Comment by FlorentBecker** [comment]
So, the problem occurs also in main(), but there is a dwarf_errno:

```
$ kcov cov/ hello
in main, dwarf_status of hello:0with errno:0and dwarf_errno:4
in DwarfParser::open, dwarf_status of hello:0with errno:0and dwarf_errno:4                                         
```

**Comment by FlorentBecker** [comment]
For what it's worth, dwarf_errno: 4 is "I/O Error".

**Comment by SimonKagstrom** [comment]
Thanks, I was just looking into that (there is a dwarf_errmsg() as well - maybe you already discovered that).

Looking at the dwarf_begin() source code:

  https://github.com/threatstack/elfutils/blob/master/libdw/dwarf_begin.c

I would guess this is caused by elf_begin() failing. Can you check with ldd which libelf is being used in both your applications?

**Comment by FlorentBecker** [comment]
```
$ ldd `which kcov`
        linux-vdso.so.1 (0x00007fff5d7b9000)
        libdw.so.1 => /nix/store/c9bi6s7mayv6blgfd2451ff2gvd9zz01-elfutils-0.163/lib/libdw.so.1 (0x00007f6d8e78e000)
        libelf.so.0 => /nix/store/nsmkdi8gxw180kqrn57risp935a9rrji-libelf-0.8.13/lib/libelf.so.0 (0x00007f6d8e577000)
        libstdc++.so.6 => /nix/store/w6xihr8fkgmmk4qqaaf5is4y4a5mm0yp-gcc-4.9.3/lib/libstdc++.so.6 (0x00007f6d8e26c000)
        libcurl.so.4 => /nix/store/mffgqkqyh2z6arii7q6i63n2ansfsgym-curl-7.45.0/lib/libcurl.so.4 (0x00007f6d8e002000)
        libopcodes-2.23.1.so => /nix/store/390z7i593kfnysfyx4sawlgxi8fvw074-binutils-2.23.1/lib64/libopcodes-2.23.1.so (0x00007f6d8dd13000)
        libbfd-2.23.1.so => /nix/store/390z7i593kfnysfyx4sawlgxi8fvw074-binutils-2.23.1/lib64/libbfd-2.23.1.so (0x00007f6d8da04000)
        libz.so.1 => /nix/store/s6zky9ysv3qpwz53xvza43n8vzijyafn-zlib-1.2.8/lib/libz.so.1 (0x00007f6d8d7ec000)
        libpthread.so.0 => /nix/store/y8scgsh8warhiy3pl39mqxyxphybcy7p-glibc-2.21/lib/libpthread.so.0 (0x00007f6d8d5cf000)
        libdl.so.2 => /nix/store/y8scgsh8warhiy3pl39mqxyxphybcy7p-glibc-2.21/lib/libdl.so.2 (0x00007f6d8d3cb000)
        libm.so.6 => /nix/store/y8scgsh8warhiy3pl39mqxyxphybcy7p-glibc-2.21/lib/libm.so.6 (0x00007f6d8d0c8000)
        libgcc_s.so.1 => /nix/store/y8scgsh8warhiy3pl39mqxyxphybcy7p-glibc-2.21/lib/libgcc_s.so.1 (0x00007f6d8ceb2000)
        libc.so.6 => /nix/store/y8scgsh8warhiy3pl39mqxyxphybcy7p-glibc-2.21/lib/libc.so.6 (0x00007f6d8cb12000)
        libelf.so.1 => /nix/store/c9bi6s7mayv6blgfd2451ff2gvd9zz01-elfutils-0.163/lib/libelf.so.1 (0x00007f6d8c8fb000)
        libbz2.so.1.0 => /nix/store/0iv2ykxshmq5gwfm8qbc4n7649algn3h-bzip2-1.0.6/lib/libbz2.so.1.0 (0x00007f6d8c6eb000)
        /nix/store/y8scgsh8warhiy3pl39mqxyxphybcy7p-glibc-2.21/lib/ld-linux-x86-64.so.2 (0x00007f6d8e9d4000)
        libssh2.so.1 => /nix/store/zwwwy0992disa9rf45j2bqnbaqg62fz6-libssh2-1.6.0/lib/libssh2.so.1 (0x00007f6d8c4c0000)
        libssl.so.1.0.0 => /nix/store/nx2yg6qj0rjzlaifyb1ayl440jxfjzna-openssl-1.0.1p/lib/libssl.so.1.0.0 (0x00007f6d8c252000)
        libcrypto.so.1.0.0 => /nix/store/nx2yg6qj0rjzlaifyb1ayl440jxfjzna-openssl-1.0.1p/lib/libcrypto.so.1.0.0 (0x00007f6d8be65000)

$ ldd dw_test 
        linux-vdso.so.1 (0x00007fffbffdd000)
        libdw.so.1 => /nix/store/c9bi6s7mayv6blgfd2451ff2gvd9zz01-elfutils-0.163/lib/libdw.so.1 (0x00007f4a76769000)
        libc.so.6 => /nix/store/y8scgsh8warhiy3pl39mqxyxphybcy7p-glibc-2.21/lib/libc.so.6 (0x00007f4a763c9000)
        libelf.so.1 => /nix/store/c9bi6s7mayv6blgfd2451ff2gvd9zz01-elfutils-0.163/lib/libelf.so.1 (0x00007f4a761b2000)
        libdl.so.2 => /nix/store/y8scgsh8warhiy3pl39mqxyxphybcy7p-glibc-2.21/lib/libdl.so.2 (0x00007f4a75fae000)
        libbz2.so.1.0 => /nix/store/0iv2ykxshmq5gwfm8qbc4n7649algn3h-bzip2-1.0.6/lib/libbz2.so.1.0 (0x00007f4a75d9e000)
        libz.so.1 => /nix/store/s6zky9ysv3qpwz53xvza43n8vzijyafn-zlib-1.2.8/lib/libz.so.1 (0x00007f4a75b86000)
        /nix/store/y8scgsh8warhiy3pl39mqxyxphybcy7p-glibc-2.21/lib/ld-linux-x86-64.so.2 (0x00007f4a769af000)
```

Both seem to use the same libelf!

**Comment by SimonKagstrom** [comment]
Not quite:
```
$ ldd `which kcov`
        libelf.so.0 => /nix/store/nsmkdi8gxw180kqrn57risp935a9rrji-libelf-0.8.13/lib/libelf.so.0 (0x00007f6d8e577000)

$ ldd dw_test 
        libelf.so.1 => /nix/store/c9bi6s7mayv6blgfd2451ff2gvd9zz01-elfutils-0.163/lib/libelf.so.1 (0x00007f4a761b2000)
```

and this is probably the root cause. Probably the problem is caused by cmake/FindLibElf.cmake in kcov, which maybe looks up the wrong libelf to link with. If I remember correctly, Debian/Ubuntu only allows one of the libelfs to be installed, so it can't make the wrong choice there, but maybe that's not the case in nixos?

**Comment by SimonKagstrom** [comment]
Sorry, found libelf.so.1 in the kcov listing as well now, but I still think this is the problem since it has both.

**Comment by FlorentBecker** [comment]
I'll look into that. It's clearly a problem with the packaging of kcov in nixos. Thanks a lot for the help!

**Comment by SimonKagstrom** [comment]
Actually, I'd say it's a kcov issue after all: It really depends on libelf
from elfutils, not libelf from libelfg0 (in Debian/Ubuntu at least), so it
should check for that.

I'm a bit unsure on how to do that detection in FindLibElf.cmake though.

**Comment by FlorentBecker** [comment]
Thanks, I've fixed the problem in nixpkgs by not depending of libelf(g0). Not sure what to do if someone manages to have both libelfg0 and elfutils on a non-nix system… They'll probably reopen a bug.

**Comment by SimonKagstrom** [comment]
OK, good to hear!

I've tried to find out how cmake can handle this type of dependency (on a specific library version), but it doesn't seem possible right now. Ubuntu/Debian simply marks elfutils-dev and libelfg0-dev as conflicting, so there I don't think it's possible to get into this situation.

I also googled a bit for the problem, but both elfutils and libelfg0 probably attracts a too small crowd for this to be easily found.
